component/prometheus/interceptor.go

```diff
@@ -2,81 +2,55 @@ package prometheus
 
 import (
 	"context"
-	"fmt"
-	"sync"
 
 	"github.com/prometheus/prometheus/model/exemplar"
 	"github.com/prometheus/prometheus/model/labels"
 	"github.com/prometheus/prometheus/model/metadata"
 	"github.com/prometheus/prometheus/storage"
 )
 
-// Intercept func allows interceptor owners to inject custom behavior.
-type Intercept func(ref storage.SeriesRef, l labels.Labels, t int64, v float64, next storage.Appender) (storage.SeriesRef, error)
-
-// Interceptor supports the concept of an appendable/appender that you can add a func to be called before
-// the values are sent to the child appendable/append.
+// Interceptor is a storage.Appendable which invokes callback functions upon
+// getting data. Interceptor should not be modified once created. All callback
+// fields are optional.
 type Interceptor struct {
-	mut sync.RWMutex
-	// intercept allows one to intercept the series before it fans out to make any changes. If labels.Labels returns nil the series is not propagated.
-	// Intercept shouuld be thread safe and can be called across appenders.
-	intercept Intercept
-	// next is where to send the next command.
-	next storage.Appendable
-
-	// ComponentID is what component this belongs to.
-	componentID string
-}
+	OnAppend         func(ref storage.SeriesRef, l labels.Labels, t int64, v float64, next storage.Appender) (storage.SeriesRef, error)
```

Comment:
Would it be safer to hide this in a constructor, and use wrapper functions to call the underlying funcs?

Reply:
Maybe, I was hesitant to introduce a constructor that took 4 parameters, especially since that list will keep growing. 

If you think it needs to be a constructor, I wouldn't be opposed to the functional option pattern like `NewInterceptor(next, WithAppendHook(fn), WithMetadataHook(fn))` though.

Reply:
Realistically it's unlikely to get misused but if the functional pattern is easy and quick to do then that avoids the issue entirely.

Reply:
I implemented the functional option pattern, WDYT?